daffodil-runtime2/src/main/resources/c/libruntime/errors.c

```diff
@@ -0,0 +1,200 @@
+        return "stack empty, stopping program";
+    case ERR_STACK_OVERFLOW:
+        return "stack overflow, stopping program";
+    case ERR_STACK_UNDERFLOW:
+        return "stack underflow, stopping program";
+    case ERR_STREAM_EOF:
+        return "EOF in stream, stopping program";
+    case ERR_STREAM_ERROR:
+        return "error in stream, stopping program";
+    case ERR_STRTOBOOL:
+        return "error converting XML data '%s' to boolean";
+    case ERR_STRTOD_ERRNO:
+        return "error converting XML data '%s' to number";
+    case ERR_STRTOI_ERRNO:
+        return "error converting XML data '%s' to integer";
+    case ERR_STRTONUM_EMPTY:
+        return "found no number in XML data '%s'";
+    case ERR_STRTONUM_NOT:
+        return "found non-number characters in XML data '%s'";
+    case ERR_STRTONUM_RANGE:
+        return "number in XML data '%s' out of range";
+    case ERR_XML_DECL:
+        return "error making new XML declaration";
+    case ERR_XML_ELEMENT:
+        return "error making new XML element '%s'";
+    case ERR_XML_ERD:
+        return "unexpected ERD typeCode %" PRId64 " while reading XML data";
+    case ERR_XML_GONE:
+        return "ran out of XML data";
+    case ERR_XML_INPUT:
+        return "unable to read XML data from input file";
+    case ERR_XML_LEFT:
+        return "did not consume all of the XML data, '%s' left";
+    case ERR_XML_MISMATCH:
+        return "found mismatch between XML data and infoset '%s'";
+    case ERR_XML_WRITE:
+        return "error writing XML document";
+    default:
+        return "unrecognized error code, shouldn't happen";
+    }
+}
+
+// print_maybe_stop - print a message and maybe stop the program
+
+static void
+print_maybe_stop(const Error *err, int status)
+{
+    const int   errnum = 0;
+    const char *format = "%s";
+    const char *msg = error_message(err->code);
+
+    switch (err->code)
+    {
+    case ERR_FILE_OPEN:
+    case ERR_FIXED_VALUE:
+    case ERR_INFOSET_READ:
+    case ERR_INFOSET_WRITE:
+    case ERR_STRTOBOOL:
+    case ERR_STRTOD_ERRNO:
+    case ERR_STRTOI_ERRNO:
+    case ERR_STRTONUM_EMPTY:
+    case ERR_STRTONUM_NOT:
+    case ERR_STRTONUM_RANGE:
+    case ERR_XML_ELEMENT:
+    case ERR_XML_LEFT:
+    case ERR_XML_MISMATCH:
+        error(status, errnum, msg, err->s);
+        break;
+    case ERR_CHOICE_KEY:
+    case ERR_PARSE_BOOL:
+    case ERR_XML_ERD:
+        error(status, errnum, msg, err->d64);
+        break;
+    default:
+        error(status, errnum, format, msg);
+        break;
+    }
+}
+
+// need_diagnostics - return pointer to validation diagnostics
+
+Diagnostics *
+need_diagnostics(void)
+{
+    static Diagnostics validati;
+    return &validati;
+}
+
+// add_diagnostic - add a new error to validation diagnostics
+
+void
+add_diagnostic(Diagnostics *validati, const Error *error)
+{
+    if (validati && error)
+    {
+        if (validati->length <
+            sizeof(validati->array) / sizeof(*validati->array))
```

Comment:
I guess the only way more than one can accumulate here is if they are warnings. 

But to continue execution when this is full feels like the wrong choice. 

Maybe return a boolean which indicates whether the diagnostic was able to be recorded or not. 

Then the place issuing the warning can decide if there are "too many warnings" and exit, or not?